handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java

```diff
@@ -0,0 +1,216 @@
+    private static final long MIN_TIMEOUT_NANOS = TimeUnit.MILLISECONDS.toNanos(1);
+
+    private final long timeoutNanos;
+    private final boolean closeOnTimeout;
+    private final boolean suppressReadAfterTimeout;
+
+    private ChannelFutureListener channelOnWriteListener;
+    private ReadTimeoutTask readTimeoutTask;
+
+    private boolean readSawAfterWrite;
+    private ScheduledFuture<?> timeout;
+    private boolean timedOut;
+
+    private volatile boolean removed;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param timeout read timeout after last write to the channel
+     * @param unit the {@link TimeUnit} of {@code timeout}
+     * @param closeOnTimeout if channel should be closed by this handler after firing {@link ReadTimeoutException}
+     * @param suppressReadAfterTimeout channelRead got after timeout before new write would be ignored if {@code true}.
+     * This address possible situation when we ran into timeout and get response in a very short time
+     * and timeout job executing before channelRead event.
+     * Also this addresses situation when we don't close connection after timeout ({@code closeOnTimeout == false}) and
+     * should skip next read which was received after timeout.
+     * User event {@link ReadSuppressed} will be thrown if case of read suppressing.
+     */
+    public ReadAfterWriteTimeoutHandler(long timeout, TimeUnit unit, boolean closeOnTimeout,
+                                        boolean suppressReadAfterTimeout) {
+        if (timeout <= 0) {
+            timeoutNanos = 0;
+        } else {
+            timeoutNanos = Math.max(unit.toNanos(timeout), MIN_TIMEOUT_NANOS);
+        }
+        this.closeOnTimeout = closeOnTimeout;
+        this.suppressReadAfterTimeout = suppressReadAfterTimeout;
+    }
+
+    /**
+     * Creates a new instance with {@code closeOnTimeout == true} and {@code ignoreReadAfterTimeout == true}
+     *
+     * @param timeout read timeout
+     * @param unit the {@link TimeUnit} of {@code timeout}
+     */
+    public ReadAfterWriteTimeoutHandler(long timeout, TimeUnit unit) {
+        this(timeout, unit, true, true);
+    }
+
+    @Override
+    public void handlerAdded(final ChannelHandlerContext ctx) throws Exception {
+        removed = false;
+
+        if (channelOnWriteListener == null) {
+            channelOnWriteListener = new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture future) throws Exception {
+                    //ensure that only the last write listener in a batch would setup one final timeout task
+                    //or nothing in a case of error
+                    disable();
+                    timedOut = false;
+                    if (future.isSuccess()) {
+                        if (removed) {
+                            return;
+                        }
+                        readSawAfterWrite = false;
+                        if (timeoutNanos > 0) {
```

Comment:
me:

> what does this handler do if `timeoutNanos <= 0`? Should we even allow this?

@Spikhalskiy:

> ```
if (timeout <= 0) {
            timeoutNanos = 0;
        } else {
            timeoutNanos = Math.max(unit.toNanos(timeout), MIN_TIMEOUT_NANOS);
        }
```

> This is done in a consistent manner with the code in other handlers in ```io.netty.handler.timeout```

Yes I see the guard against negative in the constructor, but it doesn't look like the handler does anything meaningful if `timeout <= 0` (has this code changed or am I missing something?). I don't think there is much value in having a `ReadTimeoutHandler` with a read timeout `<=0` either so I wouldn't worry about being consistent here.